example_plugins/src/services/labels_service.py

```
@@ -0,0 +1,115 @@
+        4. Yields control to the caller (LabelsProvider)
+        5. The caller modifies the labels IN PLACE
+        6. On exit, writes the modified labels and commits the transaction
+
+        The key insight: The caller modifies the yielded labels object directly,
+        and this context manager persists those changes atomically.
+
+        For systems that don't need locking (e.g., in-memory stores), this can
+        be simplified to:
+        ```py
+        labels = self.read_labels(entity)
+        yield labels
+        # write the labels here
+        """
+        entity_key = str(entity)
+
+        with scoped_session(commit=False, database=self.database) as session:
+            try:
+                # Use SELECT FOR UPDATE to acquire a row-level lock
+                stmt = select(EntityLabelsModel).where(EntityLabelsModel.entity_key == entity_key).with_for_update()
+                result = session.execute(stmt).scalars().first()
+
+                if result is None:
+                    labels = EntityLabels()
+                else:
+                    labels = EntityLabels.deserialize(result.labels)
+
+                # Yield control - The default LabelsProvider will modify the labels IN PLACE
+                yield labels
+
+                # After yield, write the modified labels back
+                labels_dict = labels.serialize()
+                upsert_stmt = insert(EntityLabelsModel).values(entity_key=entity_key, labels=labels_dict)
+                upsert_stmt = upsert_stmt.on_conflict_do_update(index_elements=['entity_key'], set_=labels_dict)
+                session.execute(upsert_stmt)
+
+                session.commit()
+                logger.debug(f'Committed atomic read-modify-write for entity {entity_key}')
+
+            except Exception:
+                session.rollback()
+                logger.error(f'Rolled back atomic read-modify-write for entity {entity_key}')
+                raise
```

Should raise the exception received instead of blank (unless python does that automagically?)